### PR TITLE
Replaced WebUtility.HtmlEnode with IHtmlEncoder.HtmlEncode

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #2012 @yishaigalatzer @pranavkm

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 1]
```diff
@@ -107,7 +111,7 @@ public TagBuilder GetFormInputElement([NotNull] HttpContext httpContext)
             SaveCookieTokenAndHeader(httpContext, newCookieToken);
 
             // <input type="hidden" name="__AntiForgeryToken" value="..." />
-            var retVal = new TagBuilder("input");
+            var retVal = new TagBuilder("input", _htmlEncoder);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: Can we give this a better name?

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultDisplayTemplates.cs [hunk 2]
```diff
@@ -25,12 +26,12 @@ public static string BooleanTemplate(IHtmlHelper htmlHelper)
 
             return htmlHelper.ViewData.ModelMetadata.IsNullableValueType ?
                 BooleanTemplateDropDownList(htmlHelper, value) :
-                BooleanTemplateCheckbox(value ?? false);
+                BooleanTemplateCheckbox(value ?? false, htmlHelper.HtmlEncoder);
         }
 
-        private static string BooleanTemplateCheckbox(bool value)
+        private static string BooleanTemplateCheckbox(bool value, IHtmlEncoder htmlEncoder)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultDisplayTemplates.cs]: This will work, but it doesn't seem great to thread this service through all of HtmlHelper's code. let's try and find a better way.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs [hunk 3]
```diff
@@ -106,6 +113,12 @@ public ViewDataDictionary ViewData
             }
         }
 
+        public IHtmlEncoder HtmlEncoder { get; }
+
+        public IUrlEncoder UrlEncoder { get; }
+
+        public IJavaScriptStringEncoder JavaScriptStringEncoder { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs]: docs

File: test/Microsoft.AspNet.Mvc.Core.Test/AntiXsrf/AntiForgeryWorkerTest.cs [hunk 4]
```diff
@@ -94,7 +97,8 @@ public void ChecksSSL_GetFormInputElement_Throws()
                 serializer: null,
                 tokenStore: null,
                 generator: null,
-                validator: null);
+                validator: null,
+                htmlEncoder: null);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/AntiXsrf/AntiForgeryWorkerTest.cs]: Might want to just fill these with `new HtmlEncoder()` while we're here. Eventually this will have to pass in something  non-null.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultDisplayTemplates.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultDisplayTemplates.cs]: Looked at this in more detail and it would be a pretty substantial refactor. Might we worth considering at some point but not as part of this change.

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultDisplayTemplates.cs]: not a substantial refactor since the template `Func`s all have the `IHtmlHelper` parameter.  make sure the necessary encoder is a prop on `IHtmlHelper` and use that instead.  BTW it's useless to make anything `public` on `HtmlHelper` unless it's also in the interface.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs [hunk 5]
```diff
@@ -4,18 +4,19 @@
...
+using Microsoft.Framework.WebEncoders;
 
 namespace Microsoft.AspNet.Mvc.Rendering
 {
     public class TagBuilder
     {
         private string _innerHtml;
+        private IHtmlEncoder _htmlEncoder;
 
-        public TagBuilder(string tagName)
```
Comment (GrabYourPitchforks) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: Should we keep the `TagBuilder(string tagName)` ctor overload for convenience? I'm sure there's plenty of existing code that uses this, and it's a bad experience to force devs to provide it every time.

You could also imagine a `HtmlHelper.CreateTagBuilder(string tagName)` convenience factory that instantiates a TagBuilder with the given name and with any services that the HtmlHelper is able to provide. This allows you to add ctor overloads in the future, and anybody who is calling the CreateTagBuilder convenience method would just get the correct behaviors automatically.

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 6]
```diff
@@ -264,7 +267,7 @@ public virtual void WriteTo(TextWriter writer, string value)
         {
             if (!string.IsNullOrEmpty(value))
             {
-                writer.Write(WebUtility.HtmlEncode(value));
+                writer.Write(HtmlEncoder.HtmlEncode(value));
```
Comment (GrabYourPitchforks) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: Use `HtmlEncoder.HtmlEncode(value, writer)` instead. That way the encoder writes directly to the backing TextWriter.

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultDisplayTemplatesTest.cs [hunk 7]
```diff
@@ -29,17 +29,17 @@ public class DefaultDisplayTemplatesTest
                     { "<blink>text</blink>", false, "<blink>text</blink>" },
                     { "<blink>text</blink>", true, "&lt;blink&gt;text&lt;/blink&gt;" },
                     { "&'\"", false, "&'\"" },
-                    { "&'\"", true, "&amp;&#39;&quot;" },
```
Comment (GrabYourPitchforks) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultDisplayTemplatesTest.cs]: If you're unit-testing, don't rely on the default implementation of the encoders, as they're subject to change. Consider instead using a test encoder, such as the below.

File: test/Microsoft.AspNet.Mvc.Razor.Test/RazorViewTest.cs [hunk 8]
```diff
@@ -359,14 +363,11 @@ public class RazorViewTest
         public async Task RenderAsync_ExecutesLayoutPages()
         {
             // Arrange
-            var expected =
-@"layout-content
-head-content
-body-content
-foot-content";
+            var expected = @"layout-content&#xD;&#xA;head-content&#xD;&#xA;body-content&#xD;&#xA;foot-content";
```
Comment (GrabYourPitchforks) [inline test/Microsoft.AspNet.Mvc.Razor.Test/RazorViewTest.cs]: Were these newline characters supposed to be written as literals instead of being run through the encoder?

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs [hunk 9]
```diff
@@ -117,7 +119,7 @@ private void AppendAttributes(StringBuilder sb)
                     continue;
                 }
 
-                var value = WebUtility.HtmlEncode(attribute.Value);
+                var value = _htmlEncoder.HtmlEncode(attribute.Value);
```
Comment (GrabYourPitchforks) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: Consider calling the overload that takes a TextWriter and refactoring the containing method to use StringWriter instead of StringBuilder.

**Comment by GrabYourPitchforks** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs (hunk 5 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: https://github.com/aspnet/Mvc/pull/2076/files#diff-8ab7a450491697c2c67527de5c072a8fR19

**Comment by dougbu** [comment]
@GrabYourPitchforks FYI @ajaybhargavb updated this PR.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultDisplayTemplates.cs [hunk 10]
```diff
@@ -10,6 +10,7 @@
 using Microsoft.AspNet.Mvc.Core;
 using Microsoft.AspNet.Mvc.ModelBinding;
 using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.WebEncoders;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultDisplayTemplates.cs]: why is this needed?  don't see any `IHtmlEncoder` uses

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultDisplayTemplates.cs]: Oops. Forgot to remove it.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs [hunk 11]
```diff
@@ -107,6 +114,15 @@ public ViewDataDictionary ViewData
...
+        public IHtmlEncoder HtmlEncoder { get; }
+
+        /// <inheritdoc />
+        public IUrlEncoder UrlEncoder { get; }
+
+        /// <inheritdoc />
+        public IJavaScriptStringEncoder JavaScriptStringEncoder { get; }
+
+        /// <inheritdoc />
         public IModelMetadataProvider MetadataProvider { get; private set; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs]: nit: don't think anything other than the constructor sets this property.  if I'm correct, please remove `private set;` while you're in neighbouring code.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs [hunk 12]
```diff
@@ -4,18 +4,23 @@
...
 
 namespace Microsoft.AspNet.Mvc.Rendering
 {
     public class TagBuilder
     {
         private string _innerHtml;
+        private IHtmlEncoder _htmlEncoder;
 
-        public TagBuilder(string tagName)
+        public TagBuilder(string tagName) : this(tagName, HtmlEncoder.Default)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: nit: move `this(tagName, HtmlEncoder.Default)` to a new line

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs [hunk 13]
```diff
@@ -4,18 +4,23 @@
...
+        {
+        }
+
+        public TagBuilder(string tagName, IHtmlEncoder htmlEncoder)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: `[NotNull]` for the `htmlEncoder` parameter

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs (hunk 9 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: @ajaybhargavb why did you choose not to change `ToString()` to use a `StringWriter` instead?

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: Didn't address this comment yet. Had some weird build issues. So updated the PR with rest of the fixes.

File: test/Microsoft.AspNet.Mvc.Razor.Test/RazorViewTest.cs [hunk 14]
```diff
@@ -165,17 +167,19 @@ public class RazorViewTest
         public async Task RenderAsync_AsPartial_ExecutesLayout_ButNotViewStartPages()
         {
             // Arrange
-            var expected = string.Join(Environment.NewLine,
+            var expected = string.Join("&#xD;&#xA;",
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Razor.Test/RazorViewTest.cs]: this won't work on non-Windows machines.  use a stable separation string

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultDisplayTemplatesTest.cs (hunk 7 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultDisplayTemplatesTest.cs]: @ajaybhargavb if you're not going to switch from `new HtmlEncoder()` to `new CustomHtmlEncoder()` or `new TestHtmlEncoder()` as part of this issue, you'll need a follow-up issue.  we shouldn't be unit testing @GrabYourPitchforks's code here, despite the specifics of this particular theory data.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs (hunk 9 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: :ok: please comment when you've added (an) additional commit(s) to cover the still-open comments

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultDisplayTemplatesTest.cs (hunk 7 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultDisplayTemplatesTest.cs]: Created a follow-up issue. https://github.com/aspnet/Mvc/issues/2085

**Comment by ajaybhargavb** [comment]
Updated.

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 15]
```diff
@@ -107,11 +111,11 @@ public TagBuilder GetFormInputElement([NotNull] HttpContext httpContext)
             SaveCookieTokenAndHeader(httpContext, newCookieToken);
 
             // <input type="hidden" name="__AntiForgeryToken" value="..." />
-            var retVal = new TagBuilder("input");
-            retVal.Attributes["type"] = "hidden";
-            retVal.Attributes["name"] = _config.FormFieldName;
-            retVal.Attributes["value"] = _serializer.Serialize(formToken);
-            return retVal;
+            var inputTag = new TagBuilder("input", _htmlEncoder);
+            inputTag.Attributes["type"] = "hidden";
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: Super unrelated, but could we expose `Attributes` as an indexer on `TagBuilder`? We could do prettier initialization:

```C#
var inputTag = new TagBuilder("input", _htmlEncoder)
{
   ["type"] = "hidden",
   ...
}
```

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs [hunk 16]
```diff
@@ -106,7 +114,7 @@ public void GenerateId(string name, [NotNull] string idAttributeDotReplacement)
             }
         }
 
-        private void AppendAttributes(StringBuilder sb)
+        private void AppendAttributes(TextWriter tw)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: `TextWriter textWriter`

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs (hunk 15 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: @pranavkm not part of this PR and conceptually poor because the `TagBuilder` is not its attributes.  anyhow, if you want pretty, do
``` c#
  var inputTag = new TagBuilder("input", _htmlEncoder)
  {
    Attributes =
    {
      ["type"] = "hidden",
      ...
    },
  };
```

@ajaybhargavb syntax choice here is up to you.  if you choose to use init syntax, also completely up to you whether to use `["type"] = "hidden",` or `{ "type", "hidden" },`.  it's all the same.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs (hunk 16 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: sorry, missed this when I commented on the last commit

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs [hunk 17]
```diff
@@ -179,39 +184,41 @@ public override string ToString()
...
-                    sb.Append('>')
-                        .Append(InnerHtml)
-                        .Append("</")
-                        .Append(TagName)
-                        .Append('>');
-                    break;
-            }
-
-            return sb.ToString();
+            using (var sw = new StringWriter())
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: moving up from a comment on a commit: `stringWriter`

File: src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs [hunk 18]
```diff
@@ -46,6 +47,21 @@ public interface IHtmlHelper
         ViewDataDictionary ViewData { get; }
 
         /// <summary>
+        /// Gets the <see cref="IHtmlEncoder"/> to be used for encoding Html.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs]: Html -> HTML

File: src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs [hunk 19]
```diff
@@ -46,6 +47,21 @@ public interface IHtmlHelper
...
+        /// </summary>
+        IHtmlEncoder HtmlEncoder { get; }
+
+        /// <summary>
+        /// Gets the <see cref="IUrlEncoder"/> to be used for encoding Url.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs]: Url -> "a URL"

File: src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs [hunk 20]
```diff
@@ -46,6 +47,21 @@ public interface IHtmlHelper
...
+        /// </summary>
+        IUrlEncoder UrlEncoder { get; }
+
+        /// <summary>
+        /// Gets the <see cref="IJavaScriptStringEncoder"/> to be used for encoding Javascript string.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs]: "Javascript string" -> "a JavaScript string" (capital s) or just "JavaScript"

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 21]
```diff
@@ -61,6 +61,9 @@ public HttpContext Context
         /// <inheritdoc />
         public bool IsPartial { get; set; }
 
+        [Activate]
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: doc comments

File: test/Microsoft.AspNet.Mvc.Razor.Test/RazorViewTest.cs [hunk 22]
```diff
@@ -669,7 +697,7 @@ body content
             var layoutPath = "~/Shared/Layout1.cshtml";
             viewEngine.Setup(p => p.FindPage(It.IsAny<ActionContext>(), layoutPath))
                        .Returns(new RazorPageResult(layoutPath, layoutPage));
-
+
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Razor.Test/RazorViewTest.cs]: nit: no need to add trailing whitespace

**Comment by dougbu** [comment]
@ajaybhargavb, get @GrabYourPitchforks' opinion on whether #2085 needs to be handled _before_ this goes in.

**Comment by dougbu** [comment]
:shipit: after @GrabYourPitchforks is happy

File: src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs (hunk 19 above)
Comment (GrabYourPitchforks) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs]: Take a look at the doc comments on the IXyzEncoder classes themselves if you need inspiration.

**Comment by GrabYourPitchforks** [comment]
#2085 needs to be handled eventually (preferably sooner rather than later), but I wouldn't hold up this PR for it.

**Comment by ajaybhargavb** [comment]
Updated. @NTaylorMullen

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs [hunk 23]
```diff
@@ -4,18 +4,25 @@
...
 using Microsoft.AspNet.Mvc.Core;
 using Microsoft.Framework.Internal;
+using Microsoft.Framework.WebEncoders;
 
 namespace Microsoft.AspNet.Mvc.Rendering
 {
     public class TagBuilder
     {
         private string _innerHtml;
+        private IHtmlEncoder _htmlEncoder;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: readonly

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs [hunk 24]
```diff
@@ -117,12 +125,9 @@ private void AppendAttributes(StringBuilder sb)
                     continue;
                 }
 
-                var value = WebUtility.HtmlEncode(attribute.Value);
-                sb.Append(' ')
-                    .Append(key)
-                    .Append("=\"")
-                    .Append(value)
-                    .Append('"');
+                textWriter.Write(' ' + key + "=\"");
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TagBuilder.cs]: shouldn't be doing concats like this, just use separate writes.

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs [hunk 25]
```diff
@@ -100,6 +100,8 @@ public MvcRazorHost(ICodeTreeCache codeTreeCache)
                     CreateTagHelperMethodName = "CreateTagHelper",
                     StartWritingScopeMethodName = "StartWritingScope",
                     EndWritingScopeMethodName = "EndWritingScope",
+
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: nit: delete line

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 26]
```diff
@@ -61,6 +61,12 @@ public HttpContext Context
         /// <inheritdoc />
         public bool IsPartial { get; set; }
 
+        /// <summary>
+        /// Gets the <see cref="IHtmlEncoder"/> to be used for encoding HTML.
+        /// </summary>
+        [Activate]
+        public IHtmlEncoder HtmlEncoder { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: Why do you need this? Why not just use the auto injected bits like the HtmlHelper is?

File: src/Microsoft.AspNet.Mvc.TagHelpers/LinkTagHelper.cs [hunk 27]
```diff
@@ -152,6 +150,9 @@ private enum Mode
         [Activate]
         protected internal IMemoryCache Cache { get; set; }
 
+        [Activate]
+        protected internal IHtmlEncoder HtmlEncoder { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.TagHelpers/LinkTagHelper.cs]: Need to add a test validating that this encoder is the one that's used in `BuildFallbackBlock` and in `BuildGlobbedLinkTags` (same thing for `ScriptTagHelper`)

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 28]
```diff
@@ -26,6 +26,7 @@ public static IServiceCollection AddMvc([NotNull] this IServiceCollection servic
         {
             ConfigureDefaultServices(services, configuration);
             services.TryAdd(MvcServices.GetDefaultServices(configuration));
+            services.AddEncoders(configuration);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: This should be in `ConfigureDefaultServices`

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultEditorTemplatesTest.cs [hunk 29]
```diff
@@ -808,6 +809,21 @@ public ViewDataDictionary ViewData
                 get { return _innerHelper.ViewData; }
             }
 
+            public IHtmlEncoder HtmlEncoder
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultEditorTemplatesTest.cs]: Where do these properties get used? Not seeing where.

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultTemplatesUtilities.cs [hunk 30]
```diff
@@ -191,11 +192,17 @@ public static HtmlHelper<TModel> GetHtmlHelper<TModel>(TModel model, ICompositeV
...
                     provider,
-                    urlHelper);
+                    urlHelper,
+                    new HtmlEncoder());
             }
 
             // TemplateRenderer will Contextualize this transient service.
-            var innerHelper = (IHtmlHelper)new HtmlHelper(htmlGenerator, viewEngine, provider);
+            var innerHelper = (IHtmlHelper)new HtmlHelper(htmlGenerator,
+                                                          viewEngine,
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultTemplatesUtilities.cs]: nit: dont add all the extra spaces, just have all parameters lie 1 indent in as if you hit enter on the previous line (much easier to maintain in the future). Do the same for other locations in this PR.

File: test/Microsoft.AspNet.Mvc.Razor.Test/RazorViewTest.cs [hunk 31]
```diff
@@ -359,14 +364,18 @@ public class RazorViewTest
...
         {
             // Arrange
-            var expected =
-@"layout-content
-head-content
-body-content
-foot-content";
+            var htmlEncoder = new HtmlEncoder();
+            var expected = "layout-content" +
+                           htmlEncoder.HtmlEncode(Environment.NewLine) +
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.Razor.Test/RazorViewTest.cs]: save this value and re-use. For bonus points string interpolation! :smile: :

```C#
var encodedNewLine = htmlEncoder.HtmlEncode(Environment.NewLine);
var expected = $"layout-content{encodedNewLine}head-content{encodedNewLine}body-content{encodedNewLine}foot-content{encodedNewLine}";
```

**Comment by NTaylorMullen** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultEditorTemplatesTest.cs (hunk 29 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultEditorTemplatesTest.cs]: They are just part of implementing the interface `IHtmlHelper`.

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs (hunk 26 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: @NTaylorMullen the issue with auto-injection is references to the injected properties (e.g. `Html`) don't compile.  So the `WriteTo` @ajaybhargavb changed in the next difference would cause build failures.

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: `RazorPage` needs HtmlEncoder here https://github.com/aspnet/Mvc/pull/2076/files#diff-21281ad55491937a33b7eee9b50bdc81R273

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultEditorTemplatesTest.cs (hunk 29 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultEditorTemplatesTest.cs]: @NTaylorMullen this is a test `IHtmlHelper` implementation.  the properties must exist or the compiler complains.

Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/DefaultEditorTemplatesTest.cs]: :+1:

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs (hunk 26 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: I see, alright :+1:

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs [hunk 32]
```diff
@@ -100,6 +100,8 @@ public MvcRazorHost(ICodeTreeCache codeTreeCache)
...
+                    HtmlEncoderPropertyName = "Html.HtmlEncoder",
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: sorry for commenting on an individual commit.  use new `HtmlEncoder` property.

File: src/Microsoft.AspNet.Mvc.TagHelpers/LinkTagHelper.cs (hunk 27 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.TagHelpers/LinkTagHelper.cs]: Added.

**Comment by ajaybhargavb** [comment]
@NTaylorMullen  Updated.

**Comment by NTaylorMullen** [comment]
:shipit:

**Comment by ajaybhargavb** [comment]
Updated.
